Cache output hashes in Output Manager storage

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Added a `DbUnblindedOutput` wrapper struct for the wallet to cache the hash of the output. 
Converted whole wallet to use this rather than the UnblindedOutput for use internally. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
We need to cache the hash of an output, if your wallet has enough utxo's it can block while syncing data to a base node as it has to hash every utxo. This blocking call will block the entire OMS system in the wallet. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -263,14 +268,24 @@ where
+                    .fetch_spent_outputs()
+                    .await?
+                    .into_iter()
+                    .map(|v| v.convert_to_unblinded_output())
```
the convention is a function of the form `as_unblinded_output()`

SWvheerden replied:
I just replaced this with a from

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -263,14 +268,24 @@ where
+                    .fetch_unspent_outputs()
+                    .await?
+                    .into_iter()
+                    .map(|v| v.convert_to_unblinded_output())
```
as_unblinded_output()

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -263,14 +268,24 @@ where
+            OutputManagerRequest::GetSpentOutputs => {
+                let outputs = self
+                    .fetch_spent_outputs()
+                    .await?
```
you cannot use the ? in this function as it will exit the loop. you need to pass the error back across the channel

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -263,14 +268,24 @@ where
+            OutputManagerRequest::GetUnspentOutputs => {
+                let outputs = self
+                    .fetch_unspent_outputs()
+                    .await?
```
as above

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -280,10 +295,15 @@ where
+            OutputManagerRequest::GetInvalidOutputs => {
+                let outputs = self
+                    .fetch_invalid_outputs()
+                    .await?
```
as above

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -232,7 +236,8 @@ where
         match request {
             OutputManagerRequest::AddOutput(uo) => {
-                self.add_output(uo).await.map(|_| OutputManagerResponse::OutputAdded)
+                let value = DbUnblindedOutput::create_from_unblinded_output(uo, &self.factories)?;
```
I think its more inline with the Rust conventions to call this function `from_unblinded_output`

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/models.rs:
```diff
@@ -0,0 +1,75 @@
+}
+
+impl DbUnblindedOutput {
+    pub fn convert_to_unblinded_output(self) -> UnblindedOutput {
```
```suggestion
    pub fn as_unblinded_output(self) -> UnblindedOutput {
```

SWvheerden replied:
replaced with from

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/models.rs:
```diff
@@ -0,0 +1,75 @@
+        self.unblinded_output
+    }
+
+    pub fn create_from_unblinded_output(
```
```suggestion
    pub fn from_unblinded_output(
```

SWvheerden replied:
Done

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -622,6 +625,10 @@ impl TryFrom<OutputSql> for UnblindedOutput {
+        };
+        Ok(Self {
+            unblinded_output,
+            hash: Vec::new(),
```
So this should come from the DB? Just haven't done the migration yet?

SWvheerden replied:
This is only temp for this wip PR, this will come from the db.

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -906,12 +913,12 @@ mod test {
         let commitment = factory.commit_value(&key, val.into());
         let input = TransactionInput::new(OutputFeatures::default(), commitment);
-        (input, UnblindedOutput::new(val, key, None))
+        (input, DbUnblindedOutput::new(val, key, None))
```
you need to implement DbUnblindedOutput::new(...)

philipr-za reviewed (commented):
Might as well get rid of all the clippy warnings about the redundant clones that I left in there while you are busy

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -175,6 +178,7 @@ where
 
         info!(target: LOG_TARGET, "Output Manager Service started");
         loop {
+            trace!(target: LOG_TARGET, "Starting OMS Loop");
```
I think take this out now that we have found the location of the issue. I also found it wasn't really needed because each branch has its own logging. The trace at the end of the loop was usually enough to find the culprit.

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -622,7 +629,15 @@ impl TryFrom<OutputSql> for UnblindedOutput {
             },
-        })
+        };
+        let hash = if o.hash == None {
```
This will be neater if done with an `.unwrap_or(...)`

philipr-za commented on base_layer/wallet/tests/output_manager_service/service.rs:
```diff
@@ -170,8 +172,8 @@ fn sending_transaction_and_confirmation<T: Clone + OutputManagerBackend + 'stati
     );
-    runtime.block_on(oms.add_output(uo.clone())).unwrap();
-    match runtime.block_on(oms.add_output(uo)) {
+    runtime.block_on(oms.add_output(uo.clone().into())).unwrap();
```
all of these .into() instances seem unnecessary. I don't think you should leak the use of DbUnblindedOutput out of the OMS which you do by changing the make_input(...) function in the wallet utils.

philipr-za commented on base_layer/wallet/tests/support/utils.rs:
```diff
@@ -81,12 +82,18 @@ pub fn make_input<R: Rng + CryptoRng>(
     val: MicroTari,
     factory: &CommitmentFactory,
-) -> (TransactionInput, UnblindedOutput)
+) -> (TransactionInput, DbUnblindedOutput)
```
I disagree with this. The DbUnblindedOutput should only be used within the OMS but this utility function is used to test many things in the wallet as a whole so by doing this now all the tests for the other services need to use the .into() everywhere and have to include the DbUnlindedOutput implementation.

philipr-za reviewed (commented):
Realized that with this implementation anyone with a older DB will be hashing the outputs on every read. Think we should do a form of manual migration for thos old DB's.

1. On creation of the Sqlite backend: Check if there are any Nulls in the new column
2. If there are then run a query to calc and commit the hashes

Phrase it as a migration and we can remove it in a couple releases where it wont be needed anymore.

SWvheerden committed:
Added a DbUnblidedOutput wrapper struct for the wallet to chace extra data. 

Fixed clippy warnings
Added diesel migration for new hash table.

Added manual migration function

philipr-za reviewed (approved):
lgtm
